### PR TITLE
storage: log RocksDB warnings/errors/fatals

### PR DESCRIPTION
**Pull request by ajkr**
Previously we enabled RocksDB logging only when glog verbosity was set
to three. Additionally, the RocksDB logs were printed at Cockroach's
info severity, regardless of the severity reported by RocksDB.

With this PR, RocksDB warnings, errors, and fatals are logged by
default. Info messages still require verbosity set to three to be sent
to Cockroach's log. Also, we try to set the severity in the Cockroach
log to match the severity reported by RocksDB.

Test Plan: set stall limits low and watch the log messages flow:

```
$ ./cockroach start --logtostderr --insecure
...
W190402 04:40:00.579900 138 storage/engine/rocksdb.go:97  [rocksdb]
[/home/andrew/go/src/github.com/cockroachdb/cockroach/c-deps/rocksdb/db/column_family.cc:799]
[default] Stalling writes because we have 3 level-0 files rate 3044057
W190402 04:40:00.666972 17 storage/engine/rocksdb.go:97  [rocksdb]
[/home/andrew/go/src/github.com/cockroachdb/cockroach/c-deps/rocksdb/db/column_family.cc:754]
[default] Stopping writes because we have 4 level-0 files
W190402 04:40:00.985046 482 storage/engine/rocksdb.go:97  [rocksdb]
[/home/andrew/go/src/github.com/cockroachdb/cockroach/c-deps/rocksdb/db/column_family.cc:799]
[default] Stalling writes because we have 3 level-0 files rate 3044057
W190402 04:40:00.986236 138 storage/engine/rocksdb.go:97  [rocksdb]
[/home/andrew/go/src/github.com/cockroachdb/cockroach/c-deps/rocksdb/db/column_family.cc:799]
[default] Stalling writes because we have 3 level-0 files rate 1826434
W190402 04:40:01.054917 482 storage/engine/rocksdb.go:97  [rocksdb]
[/home/andrew/go/src/github.com/cockroachdb/cockroach/c-deps/rocksdb/db/column_family.cc:754]
[default] Stopping writes because we have 4 level-0 files
W190402 04:40:01.422071 17 storage/engine/rocksdb.go:97  [rocksdb]
[/home/andrew/go/src/github.com/cockroachdb/cockroach/c-deps/rocksdb/db/column_family.cc:799]
[default] Stalling writes because we have 3 level-0 files rate 1826434
...
```

Release note: None

### CODE REVIEW

File: c-deps/libroach/ccl/db.cc [hunk 1]
```diff
@@ -148,7 +148,7 @@ rocksdb::Status DBOpenHookCCL(std::shared_ptr<rocksdb::Logger> info_log, const s
 
   // We use a logger at V(0) for encryption status instead of the existing info_log.
```
Comment (ajkr) [inline c-deps/libroach/ccl/db.cc]: I need to look into this still.

**Review by tbg — COMMENTED** [review]
Thanks for doing this. I look forward to learning so much more about RocksDB from the log files in the future.

> Reviewed 8 of 8 files at r1.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/36408#-:-LbSe8bmDmYCqCUXlu68:byug1bc)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajkr)

___
*[c-deps/libroach/options.cc, line 131 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/36408#-LbSdDl5CmEmB-lGdmDQ:-LbSdDl5CmEmB-lGdmDR:b-9b7bsz) ([raw file](https://github.com/cockroachdb/cockroach/blob/79e19361f068857967c6bc4c76f4f4509c6fb571/c-deps/libroach/options.cc#L131)):*
> ```cpp
>     // However, we do not want such separation as our Go function `rocksDBLog()` handles
>     // both concerns. So, we do all the work in the function intended for log-level checking,
>     // and then expect this function to never be called.
> ```

How often is the above function called in practice (i.e. normal rocksdb operation)? I'm asking because if it's fairly often, it might make sense to split `rocksDBLog` into two methods `rocksDBV` and `rocksDBLog` which separate the log-level check and the actual logging, to avoid formatting and then throwing away the log messages at high rate.

**Comment by ajkr** [comment]
We are certainly going to learn a lot about `Stalling writes because ...`. Hopefully some other things too!

Separating `rocksDBLog` is a good idea. The calls with info severity can be quite frequent (maybe hundreds per second?) when write rate is high as it prints every deleted/created file. I don't think this will let us move any logic into `Logv(const char* format, va_list ap)`, though, because the call to `rocksDBLog` needs to know the severity.

**Review by tbg — APPROVED** [review]
> because the call to rocksDBLog needs to know the severity.

Good point. Let me know when it's ready for review again.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/36408#-:-LbSeuv9DTiaUPCgV9BQ:b-5rqnsc)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajkr)

**Comment by ajkr** [comment]
Alright, RFAL.

**Comment by ajkr** [comment]
___
*[c-deps/libroach/ccl/db.cc, line 149 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/36408#-LbR_BRVddTtZKwX7NG1-r1-149:-LbZTJV90j7azBIe946v:b-er2xkq) ([raw file](https://github.com/cockroachdb/cockroach/blob/f43973c175986144f46b45d384498c9394e8b8b8/c-deps/libroach/ccl/db.cc#L149)):*
<details><summary><i>Previously, ajkr (Andrew Kryczka) wrote…</i></summary><blockquote>

I need to look into this still.
</blockquote></details>

Done. Added `DBLogger::info_verbosity_` to preserve the old default behavior. This logger sets it to zero so all its messages are logged by default. We give RocksDB a logger with it set to three so its info messages do not get logged by default.

**Comment by ajkr** [comment]
___
*[c-deps/libroach/options.cc, line 131 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/36408#-LbSdDl5CmEmB-lGdmDQ:-LbZTICJ1exenfxsLKbS:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/79e19361f068857967c6bc4c76f4f4509c6fb571/c-deps/libroach/options.cc#L131)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

How often is the above function called in practice (i.e. normal rocksdb operation)? I'm asking because if it's fairly often, it might make sense to split `rocksDBLog` into two methods `rocksDBV` and `rocksDBLog` which separate the log-level check and the actual logging, to avoid formatting and then throwing away the log messages at high rate.
</blockquote></details>

Done.

**Review by tbg — APPROVED** [review]
:lgtm_strong: with just one nit. Please backport this to 19.1, too.

> Reviewed 7 of 7 files at r3.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/36408#-:-LbZVk_5-ufH5bTx4zbr:bec2og2)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @ajkr and @tbg)

___
*[pkg/storage/engine/rocksdb.go, line 92 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/36408#-LbZVKqV7-VPbzIo7RHm:-LbZVKqV7-VPbzIo7RHn:boywc8) ([raw file](https://github.com/cockroachdb/cockroach/blob/647b5c4c32de7e3ddcb0125b854470ac1e11b09a/pkg/storage/engine/rocksdb.go#L92)):*
> ```Go
> func rocksDBV(sevLvl C.int, infoVerbosity C.int) bool {
> 	sev := log.Severity(sevLvl)
> 	return sev == log.Severity_INFO && log.V(int32(infoVerbosity)) ||
> ```

Nice, this means we can do `vmodule=rocksdb=3` and get verbose rocks logs, even via the web ui!
___
*[pkg/storage/engine/rocksdb.go, line 103 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/36408#-LbZVbpF2Y28suM3T8N5:-LbZVbpF2Y28suM3T8N6:by8ydva) ([raw file](https://github.com/cockroachdb/cockroach/blob/647b5c4c32de7e3ddcb0125b854470ac1e11b09a/pkg/storage/engine/rocksdb.go#L103)):*
> ```Go
> 	sev := log.Severity(sevLvl)
> 	switch {
> 	case sev == log.Severity_WARNING:
> ```

nit: can't this just be

```go
switch sev {
  case log.Severity_WARNING:
  ...
}
```

**Review by petermattis — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/36408)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @ajkr and @tbg)

___
*[c-deps/libroach/options.cc, line 81 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/36408#-LbZaFb5DGJG7QxiaGH8:-LbZaFb5DGJG7QxiaGH9:b-jb322y) ([raw file](https://github.com/cockroachdb/cockroach/blob/5e9fd0d7c6d701267f46e78d7b37690139922b56/c-deps/libroach/options.cc#L81)):*
> ```cpp
>         return;
>     }
>     if (!rocksDBV(go_log_level, info_verbosity_)) {
> ```

I wonder if this were will be any performance impact to doing this. This is a cgo crossing. Did I hear correctly that there can be thousands of these per second? That might be noticeable. I surely hate having to worry about things like this.

**Comment by ajkr** [comment]
> I wonder if this were will be any performance impact to doing this. This is a cgo crossing. Did I hear correctly that there can be thousands of these per second? That might be noticeable. 

In the default case (`rocksDBV` returns false for info logs) it shouldn't be more Go calls than before. We were previously calling `rocksDBLog` for info messages and letting it do a no-op. Now we're calling `rocksDBV`, which returns false for info logs, which lets us skip the call to `rocksDBLog`.

> I surely hate having to worry about things like this.

Agreed.

**Comment by ajkr** [comment]
For warnings/errors/fatals there are now two Go calls instead of one, but if those are happening thousands of times per second we have bigger problems to worry about :).

**Comment by petermattis** [comment]
> In the default case (rocksDBV returns false for info logs) it shouldn't be more Go calls than before. We were previously calling rocksDBLog for info messages and letting it do a no-op. Now we're calling rocksDBV, which returns false for info logs, which lets us skip the call to rocksDBLog.

Ack. Were we calling through to the rocksDBLog for verbose logs before? How many verbose (DEBUG?) logs are there? Doesn't look like that is used frequently in RocksDB.

**Comment by ajkr** [comment]
Right, we were calling it for all log severities before, the most common of which is probably INFO.

**Comment by petermattis** [comment]
Ok, thanks for the explanation. My concerns are assuaged. Carry on.

**Pull request merged**